component/file_cache/file_cache_async.go

```diff
@@ -0,0 +1,331 @@
+
+		fc.asyncSignal.Lock() //Lock on mutex to sleep until there is an entry in the map
+
+		numFailed = 1        //make numFailed non-zero value so it will enter the loop
+		for numFailed != 0 { //Loop to try to service map operations
+
+			numFailed = 0
+
+			fc.fileOps.Range(func(key, value interface{}) bool {
+
+				restTime = (math.Pow(2, tries) - 1) * 100 //restTime in ms based on number of failed tries
+				log.Trace("AsyncFileCache:: async_cloud_handler : The timeout value is %f", restTime)
+				time.Sleep(time.Duration(restTime) * (time.Millisecond))
+				val, _ := fc.fileOps.Load(key)
+
+				fileOperation := val.(FileAttributes).operation
+				fileOptions := val.(FileAttributes).options
+
+				if val != nil {
+					log.Trace("AsyncFileCache:: async_cloud_handler : The key in the function call is %s and the value is %s", key, fileOptions)
+					switch {
+					case fileOperation == "DeleteDir":
+						returnVal = fc.asyncDeleteDir(fileOptions.(internal.DeleteDirOptions))
+
+					case fileOperation == "RenameDir":
+						returnVal = fc.asyncRenameDir(fileOptions.(internal.RenameDirOptions))
+
+					case fileOperation == "CreateFile":
+						returnVal = fc.asyncCreateFile(fileOptions.(internal.CreateFileOptions))
+						// fc.fileOps.Delete(key)
+
+					case fileOperation == "DeleteFile":
+						returnVal = fc.asyncDeleteFile(fileOptions.(internal.DeleteFileOptions))
+
+					case fileOperation == "FlushFile":
+						returnVal = fc.asyncFlushFile(fileOptions.(FlushFileAbstraction))
+
+					case fileOperation == "RenameFile":
+						returnVal = fc.asyncRenameFile(fileOptions.(internal.RenameFileOptions))
+
+					case fileOperation == "CreateDir":
+						returnVal = fc.asyncCreateDir(fileOptions.(internal.CreateDirOptions))
+
+					case fileOperation == "Chmod":
+						returnVal = fc.asyncChmod(fileOptions.(internal.ChmodOptions))
+
+					case fileOperation == "Chown":
+						returnVal = fc.asyncChown(fileOptions.(internal.ChownOptions))
+
+					case fileOperation == "SyncFile":
+						returnVal = fc.asyncSyncFile(fileOptions.(internal.SyncFileOptions))
+
+					case fileOperation == "ChmodAndFlush":
+						returnVal = fc.asyncChmodAndFlush(fileOptions.(internal.ChmodOptions))
+					}
+
+					log.Trace("AsyncFileCache:: async_cloud_handler: The key after the function call is %s and the value is %s", key, fileOptions)
+
+					if returnVal == nil {
+						log.Trace("AsyncFileCache:: async_cloud_handler: File name %s has just finished file operation %s", key, fileOperation)
+						tries = 0                                 //attempt was successful, reset try counter
+						_ = fc.fileOps.CompareAndDelete(key, val) //file has been serviced, remove it from map only if file op hasn't been updated
+
+					} else {
+
+						numFailed++
+						if tries < maxTries {
+
+							tries++ //failed op, increase timeout duration
+						}
+
+					}
+				}
+				return true
+			})
+		}
+
+		//check channel to end thread here
```

Comment:
Should this be uncommented?

Reply:
I tried to implement a shut down mechanism with this for the async thread, but it was very buggy. I'll remove it and leave a comment to revisit it at some point